allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,707 @@
+                                 top_span_embeddings: torch.FloatTensor,
+                                 antecedent_embeddings: torch.FloatTensor,
+                                 antecedent_offsets: torch.FloatTensor):
+        """
+        Parameters
+        ----------
+        top_span_embeddings: torch.FloatTensor
+            Embedding representations of the top spans.
+        antecedent_embeddings: torch.FloatTensor
+            Embedding representations of the antecedent spans for each top span.
+        antecedent_offsets: torch.IntTensor
+            The offsets between each top span and its antecedent spans.
+        Returns
+        -------
+        pairwise_embeddings : torch.FloatTensor
+            Embedding representation of the pair of spans to consider. This includes both the
+            original span representations, the elementwise similarity of the span
+            representations, and an embedding representation of the distance between two spans.
+            This is used as input to pairwise classification.
+        """
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        target_embeddings = top_span_embeddings.unsqueeze(2).expand_as(antecedent_embeddings)
+        similarity_embeddings = antecedent_embeddings * target_embeddings
+
+        # Shape: (1, max_antecedents, embedding_size)
+        antecedent_distance_embeddings = self._distance_embedding(util.bucket_values(antecedent_offsets))
+
+        # Shape: (1, 1, max_antecedents, embedding_size)
+        antecedent_distance_embeddings = antecedent_distance_embeddings.unsqueeze(0)
+
+        expanded_distance_embeddings_shape = (antecedent_embeddings.size(0),
+                                              antecedent_embeddings.size(1),
+                                              antecedent_embeddings.size(2),
+                                              antecedent_distance_embeddings.size(-1))
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        antecedent_distance_embeddings = antecedent_distance_embeddings.expand(*expanded_distance_embeddings_shape)
+
+        # Shape: (batch_size, num_spans_to_keep, max_antecedents, embedding_size)
+        pairwise_embeddings = torch.cat([target_embeddings,
+                                         antecedent_embeddings,
+                                         similarity_embeddings,
+                                         antecedent_distance_embeddings], -1)
+        return pairwise_embeddings
+
+    @staticmethod
+    def _compute_antecedent_gold_labels(top_span_labels: torch.IntTensor,
+                                        antecedent_labels: torch.IntTensor):
+        """
+        Parameters
+        ----------
+        top_span_labels: torch.IntTensor
+            The cluster id label for every span.
+        antecedent_labels: torch.IntTensor
+            The cluster id label for every antecedent span.
```

Comment:
What's the shape of this?  I'm not really sure what this is, so I'm not confident in the logic below (not that I think it's wrong, I just don't understand it well enough to agree that it's correct).